Add GCEClient and GCRMClient

matt0bi opened the pull request:
**What this PR does / why we need it**:

This is part of https://github.com/spotify/gordon-janitor-gcp/pull/7 and it contains only the GCP-related clients (GCEClient and GCRMClient). It also includes the pagination mixin for the http client.

**Which issue(s) this PR fixes**
<!-- optional; in `fixes #<issue number>, fixes #<issue_number>, ...` format, will close the issue(s) when PR gets merged: -->
N/A

**Special notes for your reviewer**:
I addressed all the feedback from #7:
- Removed the retry logic from the pagination mixin.
- Renamed things to be in line [with this comment](https://github.com/spotify/gordon-janitor-gcp/pull/7#discussion_r169642245).
- Simplified the GPaginatorMixin.
- Used rst for docstring links.
- Handful of typos and docstring fixes ([like this one](https://github.com/spotify/gordon-janitor-gcp/pull/7#discussion_r169479260))

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, start the release note with the string "action required: ".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

@spotify/alf

matt0bi committed:
Extract common test helpers

matt0bi committed:
Add gcp_clients module

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,228 @@
+
+    You can find the endpoint documentation
+    `here <https://cloud.google.com/resource-manager/
+    reference/rest/#rest-resource-v1projects>`__.
```
There's a second underscore at the end of the link that isn't needed.

matt0bi replied:
Discussed this IRL - there are multiple `here` link targets and Sphinx was throwing errors about this. The double underscore creates an [anonymous hyperlink](http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#anonymous-hyperlinks), which solves this issue.

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,228 @@
+
+        You can find the endpoint documentation
+        `here <https://cloud.google.com/resource-manager/
+        reference/rest/v1/projects/list>`__.
```
Same here, two underscores.

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,228 @@
+        Returns:
+            list: list of dicts of all active projects.
+        """
+        url = f'{self.BASE_URL}/v1/projects'
```
The version of the API should be configurable (i.e. upon instantiation) - maybe default to `v1` but there have been cases where I needed to use `alpha` before, too.

econchick commented on gordon_janitor_gcp/http_client.py:
```diff
@@ -169,3 +169,30 @@ def __init__(self, auth_client=None, session=None):
+
+class GPaginatorMixin:
+    """HTTP client mixin that extracts data from paginated Google APIs.
+    """
```
Nitpick - can we get this all on one line? I also think for clarity-sake, it would make more sense to say something like "paginated _responses_ from Google APIs" since the APIs themselves aren't exactly paginated. Though that probably b0rks the single liner docstring.

econchick commented on tests/unit/conftest.py:
```diff
@@ -91,6 +91,7 @@ def fake_keyfile(fake_keyfile_data, tmpdir):
     return tmp_keyfile
 
 
+@pytest.fixture
```
Should this bit be moved into the previous commit?

econchick commented on gordon_janitor_gcp/http_client.py:
```diff
@@ -169,3 +169,30 @@ def __init__(self, auth_client=None, session=None):
+class GPaginatorMixin:
+    """HTTP client mixin that extracts data from paginated Google APIs.
+    """
+    async def get_all_pages(self, url, params):
```
The name for this method doesn't sit right for me, but I don't know what's better. I'm wanting something with `list` or `batch`..., maybe `list_all`? or `batch_list`?  🤔

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,228 @@
+            ]):
+                try:
+                    instances.append(self._extract_instance_data(instance))
+                except KeyError as e:
```
Also catch for `IndexError`s

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,228 @@
+
+        responses = await self.get_all_pages(url, params)
+        instances = []
+        self._parse_rsps_into_instances(responses, instances)
```
Have this method return instances and only take responses as an arg

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,228 @@
+
+        responses = await self.get_all_pages(url, params)
+        projects = []
+        await self._parse_rsps_into_projects(responses, projects)
```
have this method return projects and only accept responses

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+    @pytest.mark.asyncio
+    async def test_list_all_active_projects_multiple_pages(
+            self, mocker, crm_one_page_rsp, patch_crm_url, get_gce_client):
+        """Client successfully retrieves multiple pages of results from API."""
```
Is there a way to assert something was called twice? With what you have, it implicitly asserts `m.get` is called twice, but wasn't sure if there's an explicit assertion that could be made. Just a thought.

matt0bi replied:
`aioresponses` doesn't provide a simple way to do this, but it's possible to get the requests the mock client received and assert on them. It's not pretty, but if it increases the confidence in this test then I guess it works.

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+
+        expected_rsp = crm_one_page_rsp['projects']
+        expected_rsp.append(page2['projects'].pop(0))
+        assert results == expected_rsp
```
Just so we don't forget, should be expected == actual

matt0bi replied:
Regexp for doing this automagically in vim: `%s/\vassert ([a-zA-Z\(\)].+) \=\= ([a-zA-Z0-9'"\[\]_-]+)/assert \2 == \1/c` :tada:

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+        # test instance filtering when filtered by tags
+        ('maxResults=10&filter=', ['bl-tag'], 1),
+        # test instance filtering when filtered by metadata
+        ('maxResults=10&filter=', {'key': 'type', 'value': 'c0ffee-b33f'}, 1),
```
Hope we don't have any vegetarians complaining 😉

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+        gce_client = get_gce_client(gcp_clients.GCEClient)
+
+        if instance_meta:
+            blacklisted_instance = \
```
For some reason, I'm really allergic to line breaks like this 😬 If you wouldn't mind humoring me, maybe you could do something like:

```python
to_copy = compute_rsp['items']['us-west1-z']['instances'][0]
blacklisted_instance = copy.deepcopy(to_copy)
```

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+            },
+        }
+
+    @pytest.mark.parametrize('query_str,instance_meta,log_call_count', [
```
Even though I don't really like multiline decorators, it seems to be the accepted pattern. So I'm going to fix my own tests to do the same.

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+
+        if instance_meta:
+            blacklisted_instance = \
+                copy.deepcopy(
```
You know, come to think of it, you might not need this if the fixture you're copying is scoped to function (e.g. `@pytest.fixture(scope="function")`, since it'd return a new value every time it's passed into a test method/function.

matt0bi replied:
True, however I'm using deepcopy in order to create a second fixture, but with different attributes to test out if blacklisting works.

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+            self, compute_rsp, patch_compute_base_url, get_gce_client, caplog):
+        """Client ignores incomplete API replies."""
+        gce_client = get_gce_client(gcp_clients.GCEClient)
+        caplog.set_level(logging.INFO)
```
Maybe we should have a set of logging fixtures, like `caplog_info`, `caplog_debug`, `caplog_warn` so we don't override `caplog`, but also don't have to set it all the time 🤔

matt0bi replied:
I'd prefer to leave this for a later PR.

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,250 @@
+        with aioresponses() as m:
+            filter_url = f'{patch_compute_base_url}maxResults=5&filter='
+            m.get(filter_url, payload=compute_rsp)
+            url_with_token = (f'{patch_compute_base_url}maxResults=5'
```
could you leverage `filter_url`? e.g:

```python
url_with_token = f'{filter_url}&pageToken=123token123'
```

econchick commented on tests/unit/test_http_client.py:
```diff
@@ -223,3 +223,36 @@ def simple_json_callback(resp):
+# Tests & fixtures the GPagintorMixin
+#####
+@pytest.fixture
+def simple_paging_client(mocker, get_gce_client):
```
Is `get_gce_client` used?

econchick commented on tests/unit/test_http_client.py:
```diff
@@ -223,3 +223,36 @@ def simple_json_callback(resp):
+        else:
+            # last page doesn't include a nextPageToken
+            return {'data': 'final page'}
+    mocker.patch.object(simple_paging_client, 'get_json', get_json)
```
I'm a little confused (and under-caffeinated) at why you're patching this when it's defined in the `TestClient` that `simple_paging_client` returns...

matt0bi replied:
Ah, good point, it was a left-over from when there were more tests for this code. Since there's just one test, I've pulled the fixture into the test itself.

matt0bi commented:
Hi @econchick, I've finished addressing your feedback, could you PTAL?

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,234 @@
+                projects.append(project)
+        return projects
+
+    async def list_all_active_projects(self, api_version='v1', page_size=1000):
```
Is there a reason you have the API version here, and not given in the `__init__`?

econchick replied:
Also, don't forget docstrings for this.

matt0bi replied:
> Is there a reason you have the API version here, and not given in the __init__?

Is there a benefit of having it in the `__init__` vs. here?

econchick replied:
I guess I see it weird that there's a class attribute of the base URL, but the "completion" of that URL is in a method rather than class instantiation. Do you anticipate needing to make calls to the same endpoint but the different version?

matt0bi replied:
Ah, gotcha! Thanks for explaining it!

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,234 @@
+
+    async def list_instances(self,
+                             project,
+                             api_version='v1',
```
Same q here re: api version

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,234 @@
+                 auth_client=None,
+                 session=None,
+                 blacklisted_tags=None,
+                 blacklisted_metadata=None):
```
Refresh my memory - what was decided upon for formatting multiline args/kwargs in a func/meth?

matt0bi replied:
I thought we said that it's a small style difference. I don't remember if we settled on an action, but did we?

econchick replied:
Yeah I don't remember either, just that we talked about it ¯\\_(ツ)_/¯

econchick commented on gordon_janitor_gcp/gcp_clients.py:
```diff
@@ -0,0 +1,234 @@
+        }
+
+    def _blacklisted_by_tag(self, instance):
+        """Check if instance has any blacklisted tags.
```
FYI no need to have docstrings for private methods :) if you want to have them (and I'd _prefer_ not to, but open to discussion) we should be consistent.

matt0bi replied:
I added them here as a way to give more context around what's happening because I found this opaque in zonextgen. If that's too much, I can rewrite these as comments.

econchick replied:
I'd hate to nitpick, but yeah I think I'd like these as regular comments...

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,256 @@
+        expected_rsp.append(page2['projects'].pop(0))
+        assert expected_rsp == results
+        # assert requests made and their sequence
+        requests = list(m.requests.keys())
```
awesome!

econchick commented on tests/unit/test_gcp_clients.py:
```diff
@@ -0,0 +1,256 @@
+        page2 = copy.deepcopy(compute_rsp)
+        page2['items']['us-west1-z']['instances'][0]['name'] = 'instance-2'
+        compute_rsp['nextPageToken'] = '123token123'
+        with aioresponses() as m:
```
Maybe assert two requests here too like you did for the list projects one above?

matt0bi replied:
Good idea :+1:

econchick commented on tests/unit/test_http_client.py:
```diff
@@ -223,3 +223,28 @@ def simple_json_callback(resp):
+            if number_of_calls < (max_pages - 1):
+                number_of_calls += 1
+                return {'data': 'data', 'nextPageToken': 'token'}
+            else:
```
You actually don't need the `else` here since you have a `return` in the `if` clause. Here's [a good lightning talk](https://www.youtube.com/watch?v=JVVMMULwR4s) from Larry Hastings about it (starts around 4m50s).

econchick reviewed (changes requested):
Just a few small things!

matt0bi commented:
Hey @econchick, thanks for feedback! I've addressed it, could you PTAL?

I also rebased off of `develop` and had to clean up the `conftest.py` file to fix some merge conflicts/duplication, which mainly involves the [auth client](https://github.com/spotify/gordon-janitor-gcp/pull/9/commits/e9dae72b9a58e05168e4734d3dd622cbaed9ed6a#diff-c2d555a913402fd4b764d59ea63a2266R114) [fixture](https://github.com/spotify/gordon-janitor-gcp/pull/9/commits/e9dae72b9a58e05168e4734d3dd622cbaed9ed6a#diff-8c469bb5eefd0a7208b67e261c9e6420R61).

econchick reviewed (approved):
:shipit:

matt0bi commented:
:rocket:
